client/dynamicplugins/registry.go

```diff
@@ -0,0 +1,321 @@
+	d.pluginsLock.Lock()
+	defer d.pluginsLock.Unlock()
+
+	if ptype == "" {
+		// This error shouldn't make it to a production cluster and is to aid
+		// developers during the development of new plugin types.
+		return errors.New("must specify plugin type to deregister")
+	}
+	if name == "" {
+		// This error shouldn't make it to a production cluster and is to aid
+		// developers during the development of new plugin types.
+		return errors.New("must specify plugin name to deregister")
+	}
+
+	pmap, ok := d.plugins[ptype]
+	if !ok {
+		// If this occurs there's a bug in the registration handler.
+		return fmt.Errorf("no plugins registered for type: %s", ptype)
+	}
+
+	info, ok := pmap[name]
+	if !ok {
+		// plugin already deregistered, don't send events or try re-deleting.
+		return nil
+	}
+	delete(pmap, name)
+
+	broadcaster := d.broadcasterForPluginType(ptype)
+	event := &PluginUpdateEvent{
+		EventType: EventTypeDeregistered,
+		Info:      info,
+	}
+	broadcaster.broadcast(event)
+
+	return nil
+}
+
+func (d *dynamicRegistry) ListPlugins(ptype string) []*PluginInfo {
+	d.pluginsLock.RLock()
+	defer d.pluginsLock.RUnlock()
+
+	pmap, ok := d.plugins[ptype]
+	if !ok {
+		return nil
+	}
+
+	plugins := make([]*PluginInfo, 0, len(pmap))
+
+	for _, info := range pmap {
+		plugins = append(plugins, info)
+	}
+
+	return plugins
+}
+
+func (d *dynamicRegistry) DispensePlugin(ptype string, name string) (interface{}, error) {
+	d.pluginsLock.Lock()
+	defer d.pluginsLock.Unlock()
+
+	if ptype == "" {
+		// This error shouldn't make it to a production cluster and is to aid
+		// developers during the development of new plugin types.
+		return nil, errors.New("must specify plugin type to deregister")
+	}
+	if name == "" {
+		// This error shouldn't make it to a production cluster and is to aid
+		// developers during the development of new plugin types.
+		return nil, errors.New("must specify plugin name to deregister")
+	}
+
+	dispenseFunc, ok := d.dispensers[ptype]
+	if !ok {
+		// This error shouldn't make it to a production cluster and is to aid
+		// developers during the development of new plugin types.
+		return nil, fmt.Errorf("no plugin dispenser found for type: %s", ptype)
+	}
+
+	pmap, ok := d.plugins[ptype]
+	if !ok {
+		return nil, fmt.Errorf("no plugins registered for type: %s", ptype)
+	}
+
+	info, ok := pmap[name]
+	if !ok {
+		return nil, fmt.Errorf("plugin %s for type %s not found", name, ptype)
+	}
+
+	return dispenseFunc(info)
+}
+
+func (d *dynamicRegistry) PluginsUpdatedCh(ctx context.Context, ptype string) <-chan *PluginUpdateEvent {
+	b := d.broadcasterForPluginType(ptype)
+	ch := b.subscribe()
+	go func() {
+		select {
+		case <-b.shutdownCh:
+			return
+		case <-ctx.Done():
+			b.unsubscribe(ch)
+		}
+	}()
+
+	return ch
+}
+
+func (d *dynamicRegistry) Shutdown() {
+	for _, b := range d.broadcasters {
+		b.shutdown()
+	}
+}
+
+type pluginEventBroadcaster struct {
```

Comment:
So AFAICT it appears the intended behavior of the broadcaster is:

* Broadcasting is uninterruptable. Broadcasters must be shutdown before subscribers.
* Broadcasts are never dropped and consistently ordered.
* Subscribers must receive until the chan is closed.
* The above implies a receiving subscriber must not call Unsubscribe. A concurrent goroutine must do that. (Which is what the little goroutine in PluginsUpdatedCh appears to be for.)

That all sounds fine, but we should probably make sure to comment PluginsUpdatedCh thorough so the caller knows the behavior of the chan (receive until closed!).